iTwin/viewer
# Desktop CRA / Viewer initialization refactoring

This is the beginning of the desktop template (and eventual app) refactor. I made the changes in the desktop-viewer-test app for validation. I still need to fix tests, docs, etc. My goal with this initial PR is to have few people review the code and pull and run the app locally to help identify issues early on. Ideally somebody could test on Windows since I am developing on a Mac (I intend to test in a VM as well). If all goes well, I will replace the current desktop template bits with these changes

To Test:

rush build 
followed by npm start in packages/apps/desktop-viewer-test

This is the first phase in order to limit major changes. 

Still TODO:

- recent files list (persistence and retrieval)
- download local briefcase via tile context menus (view remote menu option will be replaced with "download briefcase")
- support local briefcases in the tiles (show an indicator and use local when available), recent list, etc.
- theme switcher
- disable functions based on connectivity status
- add user preference to show most file recent on startup (in lieu of the start page)

Major changes as far as the base packages are concerned are mostly around initialization:

- added initialization hooks that can be used to initialize ITJS prior to rendering a viewer
- always terminate and re-initialize the statics per instance (with the exception of IModelApp and its subclasses). Due to the way that IModelHost interacts with NativeApp, it's not feasible to "restart" the app without restarting the host at this time. My expectation is that this may be possible in 3.0 depending on if we can de-couple a few more things. 

@MichaelBelousov, added you primarily to review the new hooks for your use case (see App.tsx for usage)

@StefanRetief added for review of new initialization/termination logic to ensure it does not break sandboxes
